### PR TITLE
Add Authorize/Capture for PayPalWebPayments in Demo App

### PR DESCRIPTION
**Pull request by KunJeongPark**
### Summary of changes

- Create authorize/capture on order in demo app after successful checkout 

https://github.com/paypal/paypal-ios/assets/9273272/321a510a-eea6-4ac0-ada2-2d31cf23bdac

### Checklist

~- [ ] Added a changelog entry~

### Authors
- @KunJeongPark

### CODE REVIEW

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift
Comment (sshropshire) [inline Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift]:
```suggestion
```
Good to remove the headers.

Comment (KunJeongPark) [reply Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift]: Oops I thought I got them all. Thanks Steven.

**Review by sshropshire — APPROVED** [review]
🌊

File: Demo/Demo/Models/Order.swift
Comment (jaxdesmarais) [inline Demo/Demo/Models/Order.swift]:
```suggestion
        let accountID: String
```

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalTransactionButtonsView.swift
Comment (jaxdesmarais) [inline Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalTransactionButtonsView.swift]: Should this indent be back a bit or should it be in the `.buttonStyle`? Not sure which element it applies to based on the spacing.

File: Demo/Demo/Models/Order.swift
Comment (jaxdesmarais) [inline Demo/Demo/Models/Order.swift:26]: It doesn't look like we are using this struct anywhere currently, are there plans to use this in a future PR?

Comment (KunJeongPark) [reply Demo/Demo/Models/Order.swift:26]: I was thinking about our setupToken for vault without purchase for future. We have paymentSource struct and in iOS we have hard coded Card in there so I was thinking about this option.

Comment (KunJeongPark) [reply Demo/Demo/Models/Order.swift:26]: Without this struct,  decoding compile error on response with PaymentSource because it is expecting format with PayPal field for PayPal payments

Comment (KunJeongPark) [reply Demo/Demo/Models/Order.swift:26]: I see what you meant. I wasn't sure which fields I might use for displaying results for this PR or for the vault without purchase PR. For now, I wanted to make it explicit that the PaymentSource we are using here is both for card payments and for PayPalPayment

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalTransactionButtonsView.swift
Comment (KunJeongPark) [reply Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalTransactionButtonsView.swift]: it is a modifier for the NavigationLink which is pretty much a button. My indentation is off. I will fix it.

File: Demo/Demo/Models/Order.swift
Comment (KunJeongPark) [reply Demo/Demo/Models/Order.swift]: To use codingKey to change this, I have to include every field from the response in the CodingKeys, I think?

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalTransactionButtonsView.swift
Comment (KunJeongPark) [reply Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalTransactionButtonsView.swift]: Addressed here: 293097be913c95f2564c5f0fe0805570d66a9b4e

File: Demo/Demo/Models/Order.swift
Comment (jaxdesmarais) [reply Demo/Demo/Models/Order.swift:26]: Maybe we can just keep `email_address` for now and display it somewhere similar to what we do for `lastDigits` with Card? We can always add to this when we get to the vault with purchase feature, but can wait to add to it until we decide what we actually need across iOS and Android for those views.

Comment (KunJeongPark) [reply Demo/Demo/Models/Order.swift:26]: Addressed here: e8d1cf5f0218bab1e18e3717dcf07275907e50a9

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalWebOrderCompletionResultView.swift
Comment (scannillo) [inline Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalWebOrderCompletionResultView.swift:3]: The names of these classes confuse me a bit. What is the difference b/w `PayPalWebOrderCompletionResultView` & `PayPalWebOrderCompletionView`?

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalWebOrderCompletionView.swift
Comment (scannillo) [inline Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalWebOrderCompletionView.swift:3]: This class looks pretty much the exact same as [CardPaymentOrderCompletionView.swift](https://github.com/paypal/paypal-ios/blob/main/Demo/Demo/SwiftUIComponents/CardPaymentViews/CardPaymentOrderCompletionView.swift). The views should be re-usable across payment method/ view model types.

This way if we want to update this view, we do it in 1 place versus multiple.

File: Demo/Demo/ViewModels/PayPalWebViewModel.swift
Comment (scannillo) [inline Demo/Demo/ViewModels/PayPalWebViewModel.swift:134]: Same here, this looks like a copy-paste from `CardPaymentViewModel` ([see snippet](https://github.com/paypal/paypal-ios/blob/a490ffdc891f5b5c3137749655f08bd9c7513339/Demo/Demo/ViewModels/CardPaymentViewModel.swift#L61-L99)). That duplication is usually a good sign that we should refactor.

Happy to talk through some refactor options. We could go the route of a parent class or a protocol to represent a generic `PaymentMethodViewModel` that has a set of default functionality that is re-usable across all payment methods (ex: auth, capture).

File: PayPal.xcodeproj/project.pbxproj
Comment (scannillo) [inline PayPal.xcodeproj/project.pbxproj]: Should this be included in this PR?

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalWebOrderCompletionResultView.swift
Comment (KunJeongPark) [reply Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalWebOrderCompletionResultView.swift:3]: I have this view (`PayPalOrderCompletionView`) where I carry over order approval results to second screen to display authorize/capture results (`PayPalWebOrderCompletionResultView`)
<table>
<tr>
<td>

</td>
<td>

</td>
</tr>
</table>

File: PayPal.xcodeproj/project.pbxproj
Comment (KunJeongPark) [reply PayPal.xcodeproj/project.pbxproj]: No. It snuck in there. I will get rid of it.

File: Demo/Demo/ViewModels/PayPalWebViewModel.swift
Comment (KunJeongPark) [reply Demo/Demo/ViewModels/PayPalWebViewModel.swift:134]: That is a great suggestion. I will start with PaymentMethodViewModel protocol and incorporate it into common views.

File: PayPal.xcodeproj/project.pbxproj
Comment (KunJeongPark) [reply PayPal.xcodeproj/project.pbxproj]: Addressed here: 80ac172b81e04c24f3d9062ee5ee22e934b5546f

File: Demo/Demo/ViewModels/PayPalWebViewModel.swift
Comment (KunJeongPark) [reply Demo/Demo/ViewModels/PayPalWebViewModel.swift:134]: @scannillo , I looked into making a `PaymentMethodViewModel` protocol and extension to provide common default `authorizeOrder` and `captureOrder` function implementations and I ran into issue of having to change stored properties(states) in the conforming types. Protocol extensions, as I understand, cannot mutate stored properties of conforming types. For instance:  

`self.state.authorizedOrderResponse = .loaded(order)`

I could provide in `PaymentMethodViewModel` protocol delegate methods like `didAuthorizeOrder` and `didCaptureOrder` but it would reduce duplication only slightly and introduce complexity. 
I am looking into protocol with associatedType but it doesn't look pretty imo.

I ran into similar issue with factoring out views that looks the same - `state` in each ViewModel is of different type. I would need to use associated types and also run into issue of observing different properties of states of different types for each viewModels. 

I think benefits of reducing a bit of code duplication doesn't justify the complexity introduced.

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalWebOrderCompletionView.swift
Comment (KunJeongPark) [reply Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalWebOrderCompletionView.swift:3]: See comment below.

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalWebOrderCompletionResultView.swift
Comment (jaxdesmarais) [reply Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalWebOrderCompletionResultView.swift:3]: If the order completion has already taken place maybe `PayPalWebResultView` is more appropriate?

File: Demo/Demo/ViewModels/PayPalWebViewModel.swift
Comment (jaxdesmarais) [reply Demo/Demo/ViewModels/PayPalWebViewModel.swift:134]: It seems like there could be a few options in this StackOverflow that may be worth exploring: https://stackoverflow.com/questions/38885622/swift-protocol-extensions-property-default-values. Not sure if Sammy had a different approach in mind as well!

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalWebOrderCompletionResultView.swift
Comment (KunJeongPark) [reply Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalWebOrderCompletionResultView.swift:3]: There are several results in these flows. For PayPalWeb, I have order creation result, approval result and the order completion result (result of authorize/capture). So I am distinguishing them with the names.

File: Demo/Demo/ViewModels/PayPalWebViewModel.swift
Comment (KunJeongPark) [reply Demo/Demo/ViewModels/PayPalWebViewModel.swift:134]: I will continue working w PaymentMethodViewModel protocol. I guess all both flows have views for order creation result, order approval result and order capture/authorize result.

Comment (KunJeongPark) [reply Demo/Demo/ViewModels/PayPalWebViewModel.swift:134]: I created another ticket for refactoring common views and functionalities in the viewModels. https://paypal.atlassian.net/browse/DTPPCPSDK-1437

File: Demo/Demo/SwiftUIComponents/CardPaymentViews/OrderCreateCardResultView.swift
Comment (jaxdesmarais) [inline Demo/Demo/SwiftUIComponents/CardPaymentViews/OrderCreateCardResultView.swift:3]: In the refactor ticket can we please include updating some of these names? This name doesn't seem to match up with what the code in this file is doing and it's not clear why it's being changed since there are no other code changes in this file. I don't think the original name was super clear but it's especially confusing we made this change with no other changes here.

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift
Comment (jaxdesmarais) [inline Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift]: Since intent is an enum can we use a switch here instead?

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift
Comment (jaxdesmarais) [inline Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift]: It seems like both this and the else have basically the same code repeated. Can we pass the intent to the view model to dry up this code? Something like:
```swift
try await paypalWebViewModel(with: intent, orderID: orderID, selectedMerchantIntegration: selectedMerchantIntegration)
```
Then in the view model we can do a switch on the intent passed in.

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalTransactionsView.swift
Comment (jaxdesmarais) [inline Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalTransactionsView.swift]:
```suggestion
struct PayPalTransactionView: View {
```

Comment (KunJeongPark) [reply Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalTransactionsView.swift]: There are three options of transactions on that view PayPal, PayPal Credit, Pay Later. That's why I put Transactions. What do you think?

File: Demo/Demo/SwiftUIComponents/CardPaymentViews/OrderCreateCardResultView.swift
Comment (KunJeongPark) [reply Demo/Demo/SwiftUIComponents/CardPaymentViews/OrderCreateCardResultView.swift:3]: Basically we have result views for each Orders API calls to check if they are working as expected.
so there are OrderCreateCardResultView, OrderCreatePayPalWebResultView, etc.
And there are also container views since I didn't want them to be all on one scrollable page.

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift
Comment (KunJeongPark) [reply Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift]: I just used if since there are only two cases. I don't anticipate that there will be other forms of order operations after order approval but I am not tied to if-then.

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift
Comment (KunJeongPark) [reply Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift]: I have the intent toggle on the main screen for PayPalWebPayment.

 It would not make sense imo to instantiate the PayPalWebViewModel with intent option set. The pattern we have set is instantiating the viewModels as @StateObject with the main view for the payment method. We publish modifications in viewModel through actions from each API calls after.

Comment (jaxdesmarais) [reply Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift]: We aren't instantiating the view model here we are calling a method called `paypalWebViewModel.captureOrder/authroizeOrder`. The suggestion is to combine the logic here into 1 method on  `paypalWebViewModel`. It can be called whatever we want, so if order is more clear we can do `try await paypalWebViewModel.order(with: intent, orderID: orderID, selectedMerchantIntegration: selectedMerchantIntegration)`. I just removed the method name `captureOrder/authroizeOrder` to illustrate the idea.

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalTransactionsView.swift
Comment (jaxdesmarais) [reply Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalTransactionsView.swift]: I think the singular is fine even though it's used in multiple places there will ever only be 1 transaction

File: Demo/Demo/SwiftUIComponents/CardPaymentViews/OrderCreateCardResultView.swift
Comment (jaxdesmarais) [reply Demo/Demo/SwiftUIComponents/CardPaymentViews/OrderCreateCardResultView.swift:3]: It sounds like this will be a great thing to refactor or can be shared in some way across payment methods since most of the logic is the same. I think that will help with naming as then we can have a singular `OrderResultView`.

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift
Comment (KunJeongPark) [reply Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift]: That makes sense. I will have one function that redirects from the viewModel. Thank you!

File: Demo/Demo/SwiftUIComponents/CardPaymentViews/OrderCreateCardResultView.swift
Comment (KunJeongPark) [reply Demo/Demo/SwiftUIComponents/CardPaymentViews/OrderCreateCardResultView.swift:3]: That is what I want to investigate in the refactor ticket. How to organize the different states to minimize views where it makes sense. Most of the result views have different content. There are two that are identical, order create result views and the "action buttons" for authorize/capture. To me, it's easier to follow if there are separate views for each API call result so we can find them more easily but I will explore different options.

That's why I wanted to create a different PR for the refactor as it will be quite a bit of code change in the viewModels and some backtracking and modifications based on feedback. I wanted to be able to track these changes.

File: Demo/Demo/ViewModels/PayPalWebViewModel.swift
Comment (jaxdesmarais) [inline Demo/Demo/ViewModels/PayPalWebViewModel.swift]:
```suggestion
```

File: Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift
Comment (jaxdesmarais) [inline Demo/Demo/SwiftUIComponents/PayPalWebViews/PayPalOrderActionButton.swift:33]: 🎉

**Review by jaxdesmarais — APPROVED** [review]
Thanks for addressing all of my questions/suggestions!

File: Demo/Demo/ViewModels/PayPalWebViewModel.swift
Comment (KunJeongPark) [inline Demo/Demo/ViewModels/PayPalWebViewModel.swift:1]: Thank you!

**Pull request merged**
